### DISCUSSION TITLE
Is there a way to assign different styles to different Tk windows?

### DISCUSSION THREAD

**Discussion by Yuankun123** [israel-dryer/ttkbootstrap#259] [Q&A]
Hello,

In the original ttk, I usually use “Style(root)” to access the style of a specific window. However, it seems that ttkbootstrap doesn’t support this. So, I wonder if there is another way to do this. 

Thank you!

**Comment by israel-dryer** [comment]
Hi. As far as I know there can be only one ttk style running on an app at a given time. However, if you know of a way to do this in regular ttk, please post a reproducible example. I'd love to see it.

In ttkbootstrap, the `Style` class is a singleton object that is instantiated at run-time, so it will always return the same instance of the Style class. You can either use `Style()` or `Style.get_instance()` to get the style object. Or, if you use the `Window` class to create your application window, the you can access style from the `style` property.

```python
import ttkbootstrap as ttk

window = ttk.Window()
window.style.theme_use("darkly")

# also
style = ttk.Style()
style = ttk.Style.get_instance()
```

https://ttkbootstrap.readthedocs.io/en/latest/api/style/style/

**Reply by Yuankun123** [reply]
Hi. I am happy to share this with you… my simple “window switch game”.

 
         from tkinter import *
         from tkinter.ttk import *


         class WindowSwitchTrick:
              def __init__(self):
                  self.window1 = None
                  self.window2 = None
                  self.display_window1()
          
              def display_window1(self):
                  if self.window1 is None:
                      self.window1 = Tk()
                      self.window1.geometry('500x300')
                      self.window1.geometry('+100+200')
                      self.window1.title('Window1 with style default')
                      s1 = Style(self.window1)
                      s1.theme_use('default')
          
                      button1_dis = Button(self.window1, text='Display window2 with style classic', command=self.display_window2)
                      button1_dis.pack()
          
                      button1_des = Button(self.window1, text='Destroy window2', command=self.destroy_window2)
                      button1_des.pack()
          
                      self.window1.mainloop()
          
              def display_window2(self):
                  if self.window2 is None:
                      self.window2 = Tk()
                      self.window2.geometry('500x300')
                      self.window2.geometry('+700+200')
                      self.window2.title('Window2 with style classic')
                      s2 = Style(self.window2)
                      s2.theme_use('classic')
          
                      button2_dis = Button(self.window2, text='Display window1 with style default', command=self.display_window1)
                      button2_dis.pack()
          
                      button2_des = Button(self.window2, text='Destroy window1', command=self.destroy_window1)
                      button2_des.pack()
          
                      self.window2.mainloop()
          
              def destroy_window1(self):
                  if self.window1 is not None:
                      self.window1.destroy()
                      self.window1 = None
          
              def destroy_window2(self):
                  if self.window2 is not None:
                      self.window2.destroy()
                      self.window2 = None


        WindowSwitchTrick()


So, I believe the issue here is that, as far as I know, ttkbootstrap doesn’t support the creation of multiple windows managed by Style at the same time.

**Reply by israel-dryer** [reply]
that's very cool. This gives me something to work with over the weekend. :-)

Yes, currently, you can create multiple windows, but there is only a single style manager.

The big difference between standard ttk and what I'm doing with ttkbootstrap is that the themes are actually built on-demand. With standard ttk, all of the themes are loaded at run-time. Typically this is not an issue since there are only about 5 or 6 built-in themes. 

However, because I've added so many new themes with a host of style options for each widget in each theme it very quickly exposes limitations with how tcl/tk implements theme loading. To make certain widgets look modern, I'm using images for a lot of the widget style elements. Doing this on scale, I run into memory limitations when I try to load widget styles for all themes... it's dozens of widget styles for each theme.... which grows to thousands of images in memory, and there is a limit to how many images that tcl/tk will create. So, to get around that limitation I have to keep track of what is built and execute my own style update routines to keep tk and ttk widgets updated with various style changes etc... I'm not sure if I'll be able to incorporate multi-window themes into ttkbootstrap. But, playing with this demo did show me some things I need to fix and perhaps update in the background in general.

Thanks for the input.

**Reply by Yuankun123** [reply]
I think we can add an option that confines “the scope of themes” we are going to use in an application. After all, few applications need to switch among all themes in ttkbootstrap. That may save lots of memory.

Just a suggestion… I don’t know if it would work…

**Reply by israel-dryer** [reply]
Can you describe some scenarios in which you would need to use more than one theme at the same time? It seems like a very small use case.

**Comment by Yuankun123** [comment]
In fact, what I am trying to do is not to use two different themes, but to create two separate windows with the same theme.  But, as far as I know, one style manager can only deal with one window, so I need two style managers.

On Apr 20, 2022, at 8:54 AM, Israel Dryer ***@***.***> wrote:

﻿

Can you describe some scenarios in which you would need to use more than one theme at the same time? It seems like a very small use case.

**Comment by Yuankun123** [comment]
Just know, I have worked out a way to do this. I will upload it later.

(:D


On Apr 20, 2022, at 10:01 AM, Zou Yuankun ***@***.***> wrote:

﻿ In fact, what I am trying to do is not to use two different themes, but to create two separate windows with the same theme.  But, as far as I know, one style manager can only deal with one window, so I need two style managers.

On Apr 20, 2022, at 8:54 AM, Israel Dryer ***@***.***> wrote:

﻿

Can you describe some scenarios in which you would need to use more than one theme at the same time? It seems like a very small use case.

**Reply by israel-dryer** [reply]
Just curious, is there any particular reason you use another main window instead of a `Toplevel`?

**Reply by Yuankun123** [reply]
Well, I am developing a program that arranges courses for schools. So, you can imagine that I can not put all tables into one window.

**Reply by israel-dryer** [reply]
If all you need is another window and you don't care about mixing themes, that's what a Toplevel allows you to do. It's essentially a completely new window existing running in the same event loop.
https://ttkbootstrap.readthedocs.io/en/latest/api/window/toplevel/

I don't know enough about your project to know what you need, but I would caution adding another main window for a few reason:
* you will be creating another instance of the underlying tcl interpreter which means you cannot share tk variables or widgets between the two windows. They will be running in completely different and isolated tcl instances and event loops. You could share python variables, but you will risk racing conditions.
* unless you need a completely isolated instance of your application running alongside itself, there is no benefit to creating a new isolated tk Window instead a Toplevel. The toplevel was intended for just this scenario.

<img width="439" alt="image" src="https://user-images.githubusercontent.com/42584742/164235074-0bd2e031-4657-40f3-89e7-7453c7b39d46.png">

```python
import ttkbootstrap as ttk

def open_lots_of_windows():
    for x in range(1, 11):
        ttk.Toplevel(title=f"Window #{x}", size=(500, 300))

window = ttk.Window()

btn = ttk.Button(window, text='Open a bunch of top level windows', command=open_lots_of_windows)
btn.pack(padx=20, pady=20)

window.mainloop()
```

**Reply by Yuankun123** [reply]
Oh wow. That is just what I need. Seems like I have wasted much of time on this……

Anyway, thank you very much!

**Reply by israel-dryer** [reply]
Definitely not a waste of time. It's all worthwhile learning for everyone. I was able to find some things I needed to fix anyway that are unrelated to this issue, but revealed by it. You are part of the large group of people (including myself) who have scoured the internet for way too long sifting through bad documentation and ancient commentary trying to figure out how to do something in tkinter. :-)

**Comment by Yuankun123** [comment]
OK. I have created a new fork to do this. 
The link it https://github.com/Yuankun123/ttkbootstrap
